### DISCUSSION TITLE
innerText, outerHTML, innerHTML attributes not working

### DISCUSSION THREAD

**Discussion by Mephisto090** [cefsharp/CefSharp.Dom#10] [Q&A] [answered]
Hello dear developer,

I've now dealt with Puppeteer a bit. It's working very well so far. Unfortunately, the "innerText" attribute does not work

Example (with innerText in vb.net): 
```
Element = Await devtoolsContext.QuerySelectorAllAsync("span")
For Each Elements In Element
Await Elements.GetAttributeValueAsync(Of String)("innerText")
Next
```

It doesn't return anything with innerText. 
Do you have an idea what could be the reason?

**Comment by Mephisto090** [comment, marked as answer]
Hello,

Unfortunately, I still haven't been able to find a solution. Is it a bug?

**Reply by amaitland** [reply]
Use GetPropertyValueAsync instead of GetAttributeValueAsync for innerText.

I'm planning on adding additional methods like GetInnerTextAsync for popular/common methods.

Let me know what others you'd like to see added.

**Reply by Mephisto090** [reply]
You're welcome.

The most important are

InnerText, outerText
InnerHTML, outerHTML
id, name, style, class, value
img, alt, href, type

It would also be important, some websites have special attributes. Of course, these must also be retrievable without problems (e.g.: data-operation, data-number, etc.)

If I think of anything else, I'll be happy to write it to you

**Reply by Mephisto090** [reply]
GetPropertyValueAsync worked and now outputs InnerText and co

**Reply by amaitland** [reply]
> It would also be important, some websites have special attributes. Of course, these must also be retrievable without problems (e.g.: data-operation, data-number, etc.)

You can obtain custom data attributes via `GetAttributeValueAsync`.

The methods I'm proposing to add are simply for convenience. You can already get/set any property and attribute.

**Comment by Mephisto090** [comment]
Is there actually a possibility of a combination of

`devtoolsContext.QuerySelectorAsync("#myElementId");`

and

`var divElements = await devtoolsContext.QuerySelectorAllAsync("div");`

as it was in the web browser

E.g.:
`WebBrowser1.Document.GetElementById(TagID).GetElementsByTagName(TagName)`

**Reply by amaitland** [reply]
> Is there actually a possibility of a combination of

`QuerySelectorAllAsync` is the same as calling [querySelectorAll](https://developer.mozilla.org/en-US/docs/Web/API/Element/querySelectorAll) in `Javascript`. If you are unfamilary with query selectors then I'd also suggest reading https://developer.mozilla.org/en-US/docs/Learn/CSS/Building_blocks/Selectors#Information.3A_Selectors_based_on_relationships.

In your example it would be something like:

```c#
var divElements = await devtoolsContext.QuerySelectorAllAsync("#myElementId div");
```

**Comment by Mephisto090** [comment]


**Reply by amaitland** [reply]
Please start a new discussion as this seems unrelated.
